R/render_release.R

```diff
@@ -35,109 +35,178 @@
-  pwalk(.l = list(x = to_publish_index,
-                  y = to_publish,
-                  z = yml),
-        .f = function(x, y, z) {
-          old_wd <- getwd()
-          setwd(dir = y)
-          render_book(input = x,
-                      output_dir =  path(output_new_root, z$params$language))
-          # rename html to index.html
-          yml <- read_yaml("_bookdown.yml")
-          original_name <- yml$book_filename
-          file.rename(from = path_ext_set(path(output_new_root,
-                                               z$params$language, original_name),
-                                          ext = "html"),
-                      to = path(output_new_root,
-                                z$params$language, "index.html"))
-          setwd(old_wd)
-        })
-
-  # render the protocol-specific NEWS.Rmd
-  pwalk(.l = list(x = to_publish,
-                  y = to_publish_rel),
-        .f = function(x, y) {
-          old_wd <- getwd()
-          setwd(dir = x)
-          render(input = path(x, "NEWS.Rmd"),
-                 output_format = "html_document",
-                 output_file = "index",
-                 output_dir = path(git_root, output_root, y))
-          setwd(old_wd)
-        })
-
-  # also render the repo NEWS.md?? into a file publish/index.html
-  # which then becomes the homepage??
-
+    yaml[[i]][["output"]] <- list(`rmarkdown::html_document` = "default")
+    protocol_code <- map_chr(parameters, "protocol_code")
+    relevant <- protocol_code == protocol_code[[i]]
+    news <- map_chr(
+      dirname(names(protocol_code[relevant])),
+      ~paste(tail(readLines(file.path(.x, "NEWS.Rmd")), -1), collapse = "\n")
+    )
+    if (length(news) > 1) {
+      lang <- map_chr(parameters[relevant], "language")
+      lang <- factor(
+        lang,
+        c("nl", "en"),
+        c("\n# Nederlandse versie\n", "\n# English version \n")
+      )
+      news <- paste(lang, news, collapse = "\n")
+    }
+    writeLines(
+      c("---", as.yaml(yaml[[i]]), "---", "", news),
+      "render_NEWS.Rmd"
+    )
+    target_dir <- file.path(
+      output_root, yaml[[i]][["params"]][["protocol_code"]]
+    )
+    if (dir_exists(target_dir)) {
+      dir_delete(target_dir)
+    }
+    render(
+      "render_NEWS.Rmd", output_file = "index.html", envir = new.env(),
+      output_dir = target_dir, output_format = html_document(
+        css = "css/inbo_rapport.css"
+      )
+    )
+    file_delete("render_NEWS.Rmd")
+  }
+  protocols <- dir_ls(
+    output_root, recurse = TRUE, type = "file",
+    regexp = "[0-9]{4}\\.[0-9]{2}/index.html"
+  )
+  meta <- map(
+    protocols,
+    function(x) {
+      meta <- readLines(x, n = 40)
+      meta <- str_subset(meta, pattern = "<meta name=\"protocol")
+      z <- as.list(str_replace(meta, ".*content=\"(.*?)\".*", "\\1"))
+      names(z) <- str_replace(meta, ".*name=\"protocol-(.*?)\".*", "\\1")
+      as.data.frame(z)
```

Comment:
I'm having a hard time seeing what this code does.... can you explain a bit?

Reply:
We've stored the meta information of protocols in the HTML through the [template](https://github.com/inbo/protocolhelper/blob/882952bf707d3650abcd287719c13f25fe89a03b/inst/css/gitbook.html#L5). This [section](https://github.com/inbo/protocolhelper/blob/882952bf707d3650abcd287719c13f25fe89a03b/R/render_release.R#L166) extracts that information from the individual HTML files. And then we combine it into an overview of the available protocols (and versions). Which we append to the homepage template.